### ISSUE TITLE
Official Fedora Package: input requested

### ISSUE THREAD

**Issue by bryce-carson** [MesserLab/SLiM#175]
@bhaller,

I'm going through the process of submitting SLiM to the official Fedora software repository. Unfortunately, another software project (Simple Login Manager) is packaged in Fedora (and other distributions) under the name slim. That project also uses the same capitalization in their project website (SLiM). This causes two issues (that aren't problems in Selection on Linked Mutations itself) in the packaging process:

1. SLiM is not an acceptable package name in Fedora's repositories, as a difference in case alone does not resolve a naming conflict (according to the guidelines).
2. `/usr/bin/slim` is also used by Simple Login Manager, preventing us from using it.

To have SLiM accepted into the Fedora Project's official software repository, there are some solutions proposed by Fedora's Packaging Guidelines:

1. Rename the package to solve the conflict, which I have already done (I have renamed the package to `SLiMgui`; this does not affect users of the Copr repository; installation instructions in the SLiM Manual do not need to change yet).
2. Approach the upstream developers (one or both) and request them to rename the binary to resolve the conflict or to something less conflicting. I have thought of `SLiMcli` as a potential alternative binary name, which parallels the binary name `SLiMgui`.
3. If one or neither upstream developer is unwilling to rename a binary, include a `Conflicts:` tag in the package specification (the .spec file). This prevents users from installing both packages at the same time.

The Fedora Packaging Guidelines state the third solution is not preferable and should be used only as a last resort. This allows users of Fedora to install as many potential packages as possible from the official repositories without requiring manual conflict resolution, which is difficult for the end user.

@iwamatsu's Simple Login Manager (stylized as SLiM) is, as the name implies, a login manager for Linux systems. It's been around for quite a long time, long enough that it hasn't seen any updates since 2013.

In the Fedora Packaging Guidelines, in §[Incompatible Binary Files with Conflicting Naming](https://docs.fedoraproject.org/en-US/packaging-guidelines/Conflicts/#_incompatible_binary_files_with_conflicting_naming_and_stubborn_upstreams), it's advised that a packager (myself) try to convince one or both upstream projects to rename the binary.

This is not ideal for our case, since:

- Simple Login Manager is not fully compatible with [systemd](https://systemd.io/), a fundamental [and default] part of a modern Fedora Workstation installation. However, @Hubbitus, the Fedora package maintainer of slim, and others contributing to that package have made updates to it in the package itself, so while upstream has abandoned the project it is not stale.
- Simple Login Manager has only a single binary, which would generally not be used by the end-user except in an `~/.xinitrc` file, which launches the login manager upon system boot; end users of Selection on Linked Mutations will occasionally or frequently (depending on how you look at it) use our `/usr/bin/slim` executable to run their models non-interactively.
- The SLiM (Selection on Linked Mutations) Manual has many references to `slim`, which is commonly used by end-users and would unfortunately require changing every reference in the large set of documentation we have.
- Users of the SLiM framework would need to update their scripts, systems, etc. to point to the renamed binary, if we chose to rename it. This would not affect current installations, only future installations; unfortunately, this will cause missing file(s) errors when users have not updated scripts they recieve from others or SLiM-Extras. It would be inadvisable, to say the least, to rename our binary.

The principle binary included in the package, SLiMgui, is the graphical application; we may opt to split the package into two separate packges, one for SLiMgui, and the other for "SLiMcli". In this way, we might kick the conflict down the road for the cli binary of slim, and have SLiMgui accepted into the official Fedora software repository sooner. Necessarily, this would require users who want the command-line program to build and install it from source in the meantime, or continue to use the Copr repository (which would, or could, be deprecated in the future).

I have temporarily renamed the slim binary to SLiMcli, to parallel SLiMgui, in my package definition. This hasn't been and won't be published to Copr. It's a potential solution only at this stage. For end-users, we might advise using `alias slim=SLiMcli` in `.bashrc` or `.bash_profile`; as far as I understand, `getty` is used to bootstrap a graphical Linux environment, so `.bash_profile` isn't executed and wouldn't mask Simple Login Manager, but then our renamed binary _could_ be masked from users and referred to as `slim` still.

Alternatively, we might be permitted by the Fedora Packaging Committee to use `Conflicts: slim` in our package definition, so that users cannot have both Simple Login Manager and Selection on Linked Mutations installed at the same time. However, since they do not provide similar capabilities I think they will not permit this, as the conflict is circumstantial and not fundamental.

This issue will be referenced in my self-introduction to the devel mailing list of the Fedora Project, and hopefully we can come to a nice solution that works for both those users of Simple Login Manager and the SLiM community (us!), and have SLiMgui accepted into the official repositories soon.

Additionally, I have made significant progress on the Debian & Ubuntu package, which I will also be submitting to official repositories. The package will also be named SLiMgui there, but the decision we come to here regarding the command-line binary will influence how I approach the Debian maintainers.

To summarize, as I know this issue is long and entirely prose, there are three things we can do:

- Rename the `slim` binary to `SLiMcli`. This affects end-users and is inadvisable.
- Request permission to use a `Conflicts: slim` entry in our package definition, so that SLiM and Simple Login Manager cannot be installed on the same system.
- Split SLiMgui into two packages, one containing only SLiMgui, and the other, "SLiMcli" containing the `slim` binary and the `eidos` binary, kicking the conflict down the road.

Regards,

Bryce. 🐙

**Comment by bhaller** [comment]
Hi @bryce-carson.  I can't see renaming the binary; that's not going to happen.  Requesting a Conflicts tag seems sensible; it sounds like this Simple Login Manager is not widely used, and mostly abandoned, so a conflict seems unlikely in practice, and I hope the people responsible would agree.  Renaming/splitting the packages seems confusing and, as you say, only kicks the conflict down the road.  I appreciate the work you've put in on this, but I'd say there's a fourth option: if they deny the request for the Conflicts tag, then give up on having SLiM in the official Fedora repo.  That doesn't seem like a huge deal, since the existing installer that you created seems to be working well for people, right?  It just seems silly for us to jump through hoops and bend over backwards to avoid a name conflict with abandonware from 2013 that it sounds like people don't use much any more, if at all (can we get statistics for how many people have actually installed it through the official Fedora repo in, say, the past year?).

> Additionally, I have made significant progress on the Debian & Ubuntu package, which I will also be submitting to official repositories. The package will also be named SLiMgui there...

I would strongly suggest naming the package "slim" on all platforms if possible.  "SLiMgui" is harder to type, but more to the point, many people will be installing on computing clusters and such, and won't even want SLiMgui; they will just want the slim command-line tool.  SLiMgui, as much as I love it, is gravy.  Am I misunderstanding your intent?

**Comment by bryce-carson** [comment]
Oh, I almost forgot, I changed my mind and I made the Review Request to get things moving along. It doesn't affect your decision at all. :)

[SLiMgui Package Review Request on RedHat/Fedora Bugzilla](https://bugzilla.redhat.com/show_bug.cgi?id=1953254)

**Comment by bryce-carson** [comment]
tl;dr—I like Linux. I'll let you know what is said on the mailing lists and we'll go from there. ☑️ 

> Hi @bryce-carson. I can't see renaming the binary; that's not going to happen.

🆗 Acknowledged, we won't rename the command-line binary.

> Requesting a Conflicts tag seems sensible; it sounds like this Simple Login Manager is not widely used, and mostly abandoned, so a conflict seems unlikely in practice, and I hope the people responsible would agree. Renaming/splitting the packages seems confusing and, as you say, only kicks the conflict down the road … It just seems silly for us to jump through hoops and bend over backwards to avoid a name conflict with abandonware from 2013 that it sounds like people don't use much any more, if at all (can we get statistics for how many people have actually installed it through the official Fedora repo in, say, the past year?).

I tried looking, but I wasn't able to find statistics. Looking at [Fedora Package Sources: slim](https://src.fedoraproject.org/rpms/slim), I did see that the package maintainer and others have done some work on the project. It is essentially abandonware, in a looser sense of the term, but it isn't _entirely stale_.

I did some background reading, and Simple Login Manager was made the default for GhostBSD some years ago. Looking at GhostBSD's position on the [DistroWatch](distrowatch.com) charts, it's not a particularly popular distribution, but I've heard of it before (it is the 46th most popular distribution, measured in terms of Hits per Day on that website over the last 12 months). GhostBSD's use of Simple Login Manager certainly isn't an argument for _confusing and forcing change upon_ the SLiM userbase. That's merely the only quasi-statistic I found regarding its use.

> I appreciate the work you've put in on this, but I'd say there's a fourth option: if they deny the request for the Conflicts tag, then give up on having SLiM in the official Fedora repo. That doesn't seem like a huge deal, since the existing installer that you created seems to be working well for people, right? 

Yes, Copr has been working wonderfully. There's a similar service provided by Ubuntu, so if `Conflicts: slim` is a total blocker in the official repositories of either distro, we have the third-party repositories still.

> > Additionally, I have made significant progress on the Debian & Ubuntu package, which I will also be submitting to official repositories. The package will also be named SLiMgui there...
> 
> I would strongly suggest naming the package "slim" on all platforms if possible. "SLiMgui" is harder to type, but more to the point, many people will be installing on computing clusters and such, and won't even want SLiMgui; they will just want the slim command-line tool. SLiMgui, as much as I love it, is gravy. Am I misunderstanding your intent?

I'd prefer that as well. Regarding Selection on Linked Mutations being included in the official repositories of Fedora, Debian, Ubuntu, etc. there is an IRC channel where package renaming like this is discussed; Fedora's Packaging Guidelines read:

> If neither upstream renames, we would then approach other distributions (distributions-list[at]freedesktop.org is a good place to discuss this) about renaming that can be done in all distros. That helps end users going from one distro to another to have consistency. Length of time that the projects have been around, how popular each is, and numerous other factors may play a role in this decision. Once a decision is made, we would rename the Fedora packages to match.

It might happen that we take up the name `slim` in these repositories, and Simple Login Manager is renamed to `simplelm` or something similar.

I'll let you know what happens on the mailing lists, if you're interested. The main reason I have put in the effort to packaging SLiM is to make installation of SLiMgui easier. Were SLiMgui still only available on MacOS, I wouldn't have put in the effort to package the command-line binary. It is quite easy to build, and the system administrators of many clusters offer to install software for users if they request it (which is the case on the federation of clusters I use).

A lot of my motivation in pursuing packaging, and now inclusion in official repositories, is to learn more about the software environments that comprise a distribution of Linux. If we aren't given the go-ahead on using a `Conflicts: slim` entry in the spec file, I might make the decision as a packager to split the package and retain it as `slimgui` (a little easier to type, but guidelines suggest SLiMgui as in the upstream (you!) styling). Users could be given two options, if they use Fedora:

1. Use Copr and do the following:
      _Enable the COPR repository then install using DNF._
      1. `sudo dnf copr enable bacarson/SLiM-Selection_on_Linked_Mutations`
      2. `sudo dnf install SLiM`
2. Use Gnome Software, or Discover (if you use KDE), or "your distribution’s [software store]\(appstream://.desktop)."<sup>1</sup>
      Click the link here and allow your browser to open the link and proceed to _Step ii_, or follow both steps.
      1. Open your distribution's software centre/store and search for SLiM (you can just type slim)
      2. Click install

The quotation in the second option shows something nice: `appstream://` URIs. These are setup to open desktop Software Stores (I really try to read Store as in "depository" and not "place to purchase things") to, well, install software. The end goal, really, is to make installing _SLiMgui_ as painless, and _cool_ as possible for the end-user. I'm imagining the people at the SLiM workshops being able to go to https://www.messerlab.org/slim/ and simply click this button to install it on a distribution of Linux:

Linux (RPM, or DEB) | Mac OS
---|---
![](https://www.freedesktop.org/software/appstream/docs/images/appstream-button2.svg) | ![](https://messerlab.files.wordpress.com/2016/02/icon_package_small.jpg?w=600)

The advantage to the end-user of a desktop distribution is that their package manager will update SLiMgui as it would any other software; something that is quite easy with Flatpaks (it's already built in), but I don't know "Flatpak-ing", and we'd still require end-users to enable a third-party repository called FlatHub.

Packaging is, more or less, an exercise in being a 🐧 Linux poweruser.

**Comment by bryce-carson** [comment]
> It just seems silly for us to jump through hoops and bend over backwards to avoid a name conflict with abandonware from 2013 that it sounds like people don't use much any more, if at all (can we get statistics for how many people have actually installed it through the official Fedora repo in, say, the past year?).

Arch Linux keeps package statistics in % popularity over time. % refers to the number of reports of usage among the sample (those providing data to Arch linux, not every machine or user does). Here's the package statistics for [Simple Login Manager](https://pkgstats.archlinux.de/packages/slim).

![Screenshot_2021-04-25 slim - pkgstats](https://user-images.githubusercontent.com/51974245/116004723-3f0cb280-a5c1-11eb-935f-4b3d313fd13f.png)

**Comment by bhaller** [comment]
So that's saying that about 4% of Arch users presently use Simple Login Manager?  That's not high, but not zero.  Arch is fairly fringe and fairly power-user, right?  Too bad we can't get these stats for Fedora and Debian; those would probably be more relevant.  In any case, it certainly looks like a package on the way out.  On the other hand, since SLiM is so specialized and used only by research biologists, our installed base is probably even smaller (but more enthusiastic, and growing!).  :->

**Comment by rathann** [comment]
For comparison, here's a similar statistic from Debian: ![Debian slim package popularity graph](https://qa.debian.org/cgi-bin/popcon-png?packages=slim&show_installed=on&want_legend=on&want_ticks=on&date_fmt=%25Y-%25m&beenhere=1).

**Comment by bryce-carson** [comment]
@rathann, you beat me to it! I've been playing with Debian's JSON data for their Popularity Contest (what you just shared). I've been trying to convert the data from JSON into something tidy that R would be happy with for some time now.

Here's another view from Debian of the same data (not what I've made in R, haven't figured out JSON yet 😭).

![popcon-png](https://user-images.githubusercontent.com/51974245/116052994-7027ca00-a637-11eb-836e-773128905e76.png)

**Comment by bryce-carson** [comment]
Here's the view I was working on. This is how Simple Login Manager's usage (in terms of how many reporting machines have the package installed) has changed over the past year among Debian users.

![Rplot](https://user-images.githubusercontent.com/51974245/116058072-b2073f00-a63c-11eb-82e5-2369c343058a.png)

Anyways, there are the stats that exist. The package review hasn't had a response yet, but when it does and things move along I'll follow-up. The stats don't influence us, we won't rename the binary; it's just interesting to see, I suppose.

**Comment by bhaller** [comment]
OK, thanks.  It does look like it's on a long slow slide downward.  I wonder how many of the people who still have it installed actually use it?  Lord knows I've got a bunch of software installed on my machine that I haven't used in a decade or more.  :->  @bryce-carson, is that what the "recent" line in your plot is capturing – people who have actually recently requested an install of it, versus people who just have it on their machine as legacy cruft?  What exactly is "recent" vs. "old" vs. "vote" vs. "installed"?  Anyway, this is just idle curiosity; yes, we're just waiting to hear from the package review.  Thanks!

**Comment by bryce-carson** [comment]
https://qa.debian.org/popcon-graph.php?packages=slim&show_installed=on&show_vote=on&show_old=on&show_recent=on&show_nofiles=on&want_percent=on&want_legend=on&want_ticks=on&from_date=2013-01-01&to_date=2021-12-31&hlght_date=2013&date_fmt=%25Y-%25m&beenhere=1
>installed is the sum of the four categories below:
vote is the number of people who use this package regularly
old is the number of people who installed, but don't use this package regularly
recent-ctime is the number of people who upgraded this package recently
no-files is the number of people whose entry didn't contain enough information (atime and ctime were 0).

**Comment by bryce-carson** [comment]
For clarification (it might've been missed), each of the metrics are based upon access and creation times of files.

**Comment by bryce-carson** [comment]
> The end goal, really, is to make installing _SLiMgui_ as painless, and _cool_ as possible for the end-user. I'm imagining the people at the SLiM workshops being able to go to https://www.messerlab.org/slim/ and simply click this button to install it on a distribution of Linux:
> Linux (RPM, or DEB) 	Mac OS
> ![](https://camo.githubusercontent.com/da2b01b2e33d21725456e8b97e8131a63a39f855ddf87b4a4912b37158497f29/68747470733a2f2f7777772e667265656465736b746f702e6f72672f736f6674776172652f61707073747265616d2f646f63732f696d616765732f61707073747265616d2d627574746f6e322e737667) 	![](https://camo.githubusercontent.com/94cdb5d24accd779e11e8001483d16f68d859515df716f1e815a30c21f500fdb/68747470733a2f2f6d65737365726c61622e66696c65732e776f726470726573732e636f6d2f323031362f30322f69636f6e5f7061636b6167655f736d616c6c2e6a70673f773d363030)
> 
> The advantage to the end-user of a desktop distribution is that their package manager will update SLiMgui as it would any other software; something that is quite easy with Flatpaks (it's already built in), but I don't know "Flatpak-ing", and we'd still require end-users to enable a third-party repository called FlatHub.
> 
> Packaging is, more or less, an exercise in being a penguin Linux poweruser.

### <p align="center">You hear the faint sound of a delivery truck outside—is it an Amazon package?</p>
<p align="center"><img src="https://flatpak.org/img/delivery_truck2-bb72338f.png" alt="A Flatpak delivery truck" width="300" height="200"></p>

### Some information, some rambling

In an earlier comment of mine, I mentioned that I don't know Flatpak-_ing_. I've looked into it again. For now it is cobbled together, and experimental, but I have managed to make a Flatpak for SLiMgui. What does this mean? Why is this good? What are the disadvantages and differences from an RPM or Debian package?

1. Two-click installation—at minimum—of the gui on *every major Linux distribution*. For now, the button below points to GNOME Builder (an IDE), but if you follow the steps below you can see how it would work for end users.
        1. Perform the [quick setup](https://flatpak.org/setup/) of Flatpak on your machine.
        2. Install SLiMgui by clicking the AppStream button below, and press 'Ok' to open the link.‡
        <a href="https://dl.flathub.org/repo/appstream/org.gnome.Builder.flatpakref">
<p align="center">
<img src="https://camo.githubusercontent.com/da2b01b2e33d21725456e8b97e8131a63a39f855ddf87b4a4912b37158497f29/68747470733a2f2f7777772e667265656465736b746f702e6f72672f736f6674776172652f61707073747265616d2f646f63732f696d616765732f61707073747265616d2d627574746f6e322e737667"></p>
</a>

2. Every installation of the app is the same as all others. Flatpaks are containerized! Every user has the same libraries and dependencies inside the container, thanks to the Flatpak runtimes and the SLiMgui container itself.
3. There's no need to maintain two, or more! packages for different distributions, which each have their own packaging guidelines. This isn't a shortcoming of Linux distributions, but it makes things difficult in the long run for a smaller community of software users like ours (albeit a capable one!).
4. We can claim the SLiM name! Presently, there is [no package using the name](https://flathub.org/apps/search/slim) on Flathub.org, the main website for Flatpak distribution. Hooray for no package name or binary conflicts!
5. It is *not standard practice* to include more than one runtime in a Flatpak; they're typically used for larger graphical applications, and this is the disadvantage compared to the RPM or Debian packages. This is similar to what I proposed before, where we *split* the binary packages for the gui and the command-line programs.
6. Users might not even *want* the gui; they don't have to build it from source, the `--DBUILD_SLIMGUI=ON` is required and an extra step for those who *do want* it. Users who want the `slim` & `eidos` binary can build them from source, or they could use the Conda binary package—but admittedly I'd be more proud if they used the RPM or (experimental) Debian package—or even that Nix package that exists.⸸

### The Point

I'll read through this wiki <https://github.com/flathub/flathub/wiki/App-Submission>, and I'll ask any questions or make comments as needed. However, if the MesserLab would like to distribute a Flatpak, I'm happy to help it along. For now, we've got *plenty* of packages and choices. <https://xkcd.com/927/>

‡ This will open the AppStream URI link in GNOME Software, KDE Discover, or whatever graphical software manager is present on a user's system. If they're using any of the major distributions with one of the major desktop environments (there's about six), they will have one. If not, it's not too different from using Flathub.org directly, and typically they'll know what to do themselves (<img src="https://archlinux.org/static/logos/archlinux-logo-black-90dpi.0c696e9c0d84.png" width="48" height="16">); in any case, the instructions in Step 1 will guide the user.
⸸Should I write a Guix package? <img src="https://guix.gnu.org/static/base/img/Guix.png" width="24" height="24"> More parentheses never hurt, right (it's an entire distribution written in Guile Scheme Lisp)?

**Comment by bryce-carson** [comment]
For the purpose of *checking in*, here's the YAML "manifest" I have so far. It needs a lot more work. A lot more.

```yml
app-id: org.messerlab.slimgui
runtime: org.kde.Platform
runtime-version: '5.15'
sdk: org.kde.Sdk
finish-args:
  - --socket=wayland
  - --socket=x11
  - --share=ipc
  - --device=dri
  - --socket=pulseaudio
  - --share=network
  - --filesystem=host
  - --talk-name=com.canonical.AppMenu.Registrar # required for global menu
  - --talk-name=org.freedesktop.Flatpak # allows spawning processes on the host via flatpak-spawn --host
  # - --env=PATH=/app/bin:/usr/bin:/app/jre/bin
  # - --env=CPPFLAGS=-I/app/include # required for building some Octave forge packages
  # - --env=LDFLAGS=-L/app/lib # required for building some Octave forge packages
  # - --env=OCTAVE_HOME=/app
  - --env=PKG_CONFIG_PATH=/app/lib/pkgconfig:/app/share/pkgconfig
command: SLiMgui
modules:
  - name: SLiMgui
    # buildsystem: simple
    builddir: true
    buildsystem: cmake
    config-opts:
      - -DBUILD_SLIMGUI=ON
      - --debug-output
    # build-commands:
    # #   - qmake --version
    #   - cmake . # -DBUILD_SLIMGUI=ON
    #   - make
    #   - install -D SLiMgui /app/bin/SLiMgui
      # # These two need to be added to $PATH
      # - install -D slim /app/bin/slim
      # - install -D eidos /app/bin/eidos
      # # Colourful and Symbolic Application Icons
      # - sed -i '69,166 s/"fill:#[[:alnum:]]*"/"fill:#ffffff"/' ./QtSLiM/icons/AppIcon16.svg
      # - sed -i '62,68 d' ./QtSLiM/icons/AppIcon16.svg
      # - install --mode=0644 --no-target-directory ./QtSLiM/icons/AppIcon64.svg %{buildroot}/usr/share/icons/hicolor/scalable/apps/org.messerlab.slimgui.svg
      # - install --mode=0644 --no-target-directory ./SLiM-%{version}/QtSLiM/icons/AppIcon16.svg %{buildroot}/usr/share/icons/hicolor/symbolic/apps/org.messerlab.slimgui-symbolic.svg
      
      # # slim Mimetype Icon
      # - install --mode=0644 --no-target-directory ./SLiM-%{version}/QtSLiM/icons/DocIcon.svg %{buildroot}/usr/share/icons/hicolor/scalable/mimetypes/text-slim.svg
      
      # # Desktop Entry and Appstream XMLs
      # - install --mode=0644 ./SLiM-%{version}/org.messerlab.slimgui-mime.xml --target-directory %{buildroot}/usr/share/mime/packages/
      # - desktop-file-install --mode=0644 --dir=%{buildroot}/usr/share/applications/ ./SLiM-%{version}/org.messerlab.slimgui.desktop
      # - install --mode=0644 ./SLiM-%{version}/org.messerlab.slimgui.appdata.xml --target-directory %{buildroot}/usr/share/metainfo/
    sources:
      - type: git
        url: https://github.com/MesserLab/SLiM.git
        tag: "v3.6"
        commit: 978622ebc1a0326f91d1cee35fd91e1e0f4e0c92
```

**Comment by bryce-carson** [comment]
The Flatpak will need testing, and the manifest may need some improvements before MesserLab could request a review from the administration of Flathub, but it's mostly complete; there is an issue with the mimetype icon as it is now, but I've made posts on mailing lists and commented on a long-standing open issue for the Flatpak documentation. Once that's resolved I'd welcome anyone to test the cross-distribution Flatpak package.

```yml
app-id: org.messerlab.slimgui
runtime: org.kde.Platform
runtime-version: '5.15'
sdk: org.kde.Sdk
finish-args:
  - --socket=wayland
  - --socket=x11
  - --share=ipc
  - --device=dri
  - --socket=pulseaudio
  - --share=network
  - --filesystem=host
  - --talk-name=com.canonical.AppMenu.Registrar # required for global menu
  - --talk-name=org.freedesktop.Flatpak # allows spawning processes on the host via flatpak-spawn --host
  - --env=PKG_CONFIG_PATH=/app/lib/pkgconfig:/app/share/pkgconfig
command: SLiMgui
modules:
  - name: SLiMgui
    builddir: true
    buildsystem: cmake
    config-opts:
      - -DBUILD_SLIMGUI=ON
      - --debug-output
    sources:
      - type: git
        url: https://github.com/MesserLab/SLiM.git
        tag: "v3.6"
        commit: 978622ebc1a0326f91d1cee35fd91e1e0f4e0c92
    post-install:
      # Colourful and Symbolic Application Icons
      - sed -i '69,166 s/"fill:#[[:alnum:]]*"/"fill:#ffffff"/' ../QtSLiM/icons/AppIcon16.svg
      - sed -i '62,68 d' ../QtSLiM/icons/AppIcon16.svg
      - install --mode=0644 -D --no-target-directory ../QtSLiM/icons/AppIcon64.svg /app/share/icons/hicolor/scalable/apps/org.messerlab.slimgui.svg
      - install --mode=0644 -D --no-target-directory ../QtSLiM/icons/AppIcon16.svg /app/share/icons/hicolor/symbolic/apps/org.messerlab.slimgui-symbolic.svg
      
      # slim Mimetype
      # Line 6 in original org.messerlab.slimgui-mime.xml
      # <generic-icon name="/usr/share/icons/hicolor/scalable/mimetypes/text-slim.svg"/>
      - sed -i '6 s,usr,app,' ../org.messerlab.slimgui-mime.xml
      - sed -i '6 s,mimetypes/text-slim.svg,mimetypes/org.messerlab.slimgui-text-slim.svg,' ../org.messerlab.slimgui-mime.xml 
      - install --mode=0644 -D --no-target-directory ../org.messerlab.slimgui-mime.xml /app/share/mime/packages/org.messerlab.slimgui-mime.xml
      - install --mode=0644 -D --no-target-directory ../QtSLiM/icons/DocIcon.svg /app/extra/export/share/icons/hicolor/scalable/mimetypes/org.messerlab.slimgui-text-slim.svg
      
      # Desktop Entry and Appstream XMLs
      - desktop-file-edit --remove-key=Exec ../org.messerlab.slimgui.desktop
      - desktop-file-edit --set-key=Exec --set-value='SLiMgui %f' ../org.messerlab.slimgui.desktop
      - desktop-file-install --mode=0644 --dir=/app/share/applications/ ../org.messerlab.slimgui.desktop
      - appstream-util validate-relax ../org.messerlab.slimgui.appdata.xml
      - install --mode=0644 -D ../org.messerlab.slimgui.appdata.xml --target-directory /app/share/metainfo/
```
[slimgui.zip](https://github.com/MesserLab/SLiM/files/6412064/slimgui.zip)

**Comment by bhaller** [comment]
Hi @bryce-carson.  Well, I don't really follow all this very well, but here are some thoughts:

- Obviously a simple installation solution that works on all major Linux platforms would be great.
- It being containerized seems like it has substantial pros and cons.  Pros are consistency and predictability of the build environment; presumably there would be fewer complaints that the install failed because of this or that toolchain/OpenGL/whatever issue.  Cons are (a) complexity, (b) perhaps longer install time?, (c) maybe compatibility issues.  Regarding (c), here's my concern.  Qt recommends different specific versions of Qt for different platforms – different versions of Ubuntu vs. different versions of Debian vs. whatever.  Generally, if a platform provides a pre-installed Qt as many do, it is a version of Qt that is deemed optimal for that platform, I suppose.  Overriding that with a one-size-fits-all version of Qt inside the container might not be a good idea, and might cause problems that we would be unaware of until users encountered them on a particular platform.  Similarly, I imagine that each platform has a standard "toolchain" – compiler, linker, C++ standard library, etc. – that are tested and optimized for that platform.  I imagine the container approach overrides that, and again, I'm not sure that's a good idea.
- One executable only is an unfortunate restriction, because we would really like the user to get both `slim` and SLiMgui from running the installer.  Your current installers provide that, as I understand it (?), whereas this Flatpak approach would not.  That's a very big drawback, I think.  I think virtually all users who want to install SLiMgui also want to install `slim`, and if they have to figure out building `slim` on the command line after the installer finishes, well, that just sucks.

I appreciate your work on this, but I'm not convinced that it is better than what we presently offer.  Offering multiple options *might* be useful to people, but it might just confuse them; I think there's a strong case to be made for just providing a single option for users on a given platform, which is the best option we have available.  And that maybe seems like what we have right now.

I'm out of my depth on understanding all the ramifications of these choices, and I'm really not in a position to do all the QA that would be required.  You may not realize the extent to which I am not a Linux user.  :->  I have only ever installed Ubuntu, in a virtualbox environment; I have never used any other distro, and I have never used Ubuntu outside of virtualbox.  I go for months at a time without using Linux at all.  If we were going to adopt a new distribution strategy that required QA across many platforms (which seems like it would be required for this, given the possibility of subtle Qt bugs due to the containerization, etc.), either you would have to do all of that (which I'm certainly not going to ask you to do!) or we'd have to have a volunteer army that we do not presently have.

If I understand correctly, this whole Flatpak thing is a possible alternative to (a) the installers you have already written, which seem to work well for almost all users, and (b) the possibility of an official Fedora installer, which is presently in limbo.  Is that correct?  I would vote for sticking with (a) for now, waiting and seeing whether (b) gets approved in which case maybe we switch Fedora installation over to that scheme, and (c) putting this Flatpak stuff on the back burner until such time as the concerns with it get resolved.  I don't see that the benefit outweighs the additional complexity that would have to be explained to users.  What do you think?
